src/Cardano/Prelude/HeapWords.hs

```diff
@@ -453,3 +456,35 @@ instance HeapWords e => HeapWords (V.Vector e) where
 
 heapWordsUVector :: V.U.Unbox e => Int -> V.U.Vector e -> Int
 heapWordsUVector sz a = 5 + (V.U.length a * sz) `div` wordSize
+
+instance HeapWords Natural where
+  heapWords (NatS# _)
+    -- We have
+    --
+    -- > NatS# GmpLimb#
+    -- > type GmpLimb# = Word#
+    --
+    -- so @(NatS# n)@ requires:
+    --
+    -- - one word for the header
+    -- - one word to store the unboxed word.
+    --
+    = 1 + 1
+  heapWords (NatJ# (BN# arr))
+    -- We have
+    --
+    -- > NatJ# {-# UNPACK #-} !BigNat
+    -- > data BigNat = BN# ByteArray#
+    --
+    -- so @NatJ# (BN# xs)@ requires:
+    --
+    -- - one word for the header
+    -- - one word for storing the header of @BN#@
+    -- - the number of bytes in the byte array
+    --
+    = 1 + 1 + arrWords (I# (sizeofByteArray# arr))
+    where
+      arrWords n
+        | n <= 0 = 0
+        | 0 < n = 1 + ((n - 1) `div` wordSize)
+        | otherwise = 0 -- Needed to keep avoid the non-exhaustive patters warning.
```

Comment:
You could combine the first and last match here?

Reply:
Yes. Good catch!

Reply:
>      -- - one word for storing the header of @BN#@ 
This is either not quite right, or unclear.

Note that the `BN#` is `UNAPCK`ed into the `NatJ#` so there's no extra word for the `BN#` constructor itself. The representation is that the `NatJ#` ends up containing just the `ByteArray#`.

So the overall size is:
 * 1 for the `NatJ#` heap object header
 * 1 for the single `NatJ#` field (of type `ByteArray#`)
 * 2 for the `ByteArray#` heap object (1 for header and 1 for storing the size)
 * `1 + ((n - 1) `div` wordSize)` for the `ByteArray#` content

Note that  `1 + ((n - 1) `div` wordSize)` gives the correct result `0` for `n = 0`, so you don't need the `otherwise` case.

I'd suggest adding a separate function for `ByteArray#` sizes, and have it do the size calculation, plus the two words overhead for the heap object, e.g.

```
heapWordsByteArray# ba# =
    2                            -- for the heap object header and size
  + 1 + ((n - 1) `div` wordSize) -- for the variable sized part
  where
    n = I# (sizeofByteArray# ba#)
```